glean-core/csharp/Glean/Net/BaseUploader.cs

```diff
@@ -0,0 +1,141 @@
+
+        private (string, string)[] GetHeadersFromJSONString(string documentId, string headers, Configuration config)
+        {
+            List<(string, string)> headerList = new List<(string, string)>();
+            try
+            {
+                // Parse the headers from JSON.
+                Dictionary<string, string> parsedHeaders = JsonSerializer.Deserialize<Dictionary<string, string>>(headers);
+
+                foreach (KeyValuePair<string, string> h in parsedHeaders)
+                {
+                    headerList.Add((h.Key, h.Value));
+                }
+            }
+            catch (JsonException)
+            {
+                Console.WriteLine($"Error while parsing headers for ping {documentId}");
+            }
+
+            if (config.pingTag != null)
+            {
+                headerList.Add(("X-Debug-ID", config.pingTag));
+            }
+
+            return headerList.ToArray();
+        }
+
+        /// <summary>
+        /// Signals Glean to upload pings at the next best opportunity.
+        /// </summary>
+        internal void TriggerUpload(Configuration config)
+        {
+            // TODO: must not work like this, it should work off the main thread.
+            // FOR TESTING Implement the upload worker here and call this from Glean.cs
+
+            int waitAttempts = 0;
+            int uploadFailures = 0;
+
+            while (uploadFailures < 3)
+            {
+                FfiUploadTask incomingTask = new FfiUploadTask();
+                IntPtr test = IntPtr.Zero;
+                LibGleanFFI.glean_get_upload_task(out test, config.logPings);
+
+                incomingTask = Marshal.PtrToStructure<FfiUploadTask>(test);
+
+                UploadTaskTag tag = UploadTaskTag.Upload;//(UploadTaskTag)incomingTask.tag;
+                switch (tag)
+                {
+                    case UploadTaskTag.Upload:
+                        {
+                            // Extract C#-friendly data from the FFI object
+                            /*string documentId = incomingTask.body.documentId;
+                            string path = incomingTask.body.path;
+                            (string, string)[] headers = GetHeadersFromJSONString(documentId, incomingTask.body.headers, config);
+                            byte[] body = new byte[incomingTask.body.bodyLen];
+                            Marshal.Copy(incomingTask.body.body, body, 0, body.Length);
+
+                            // Delegate the actual upload and get its return value.
+                            UploadResult result = Upload(path, body, headers, config);
+
+                            if (result is RecoverableFailure)
+                            {
+                                uploadFailures += 1;
+                            }
+
+                            // Process the upload response in the core.
+                            LibGleanFFI.glean_process_ping_upload_response(ref incomingTask, result.ToFfi());
```

Comment:
All data is read correctly without the last commit, but it crashes after executing this function, because Rust frees `incomingTask`